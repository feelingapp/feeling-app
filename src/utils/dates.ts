import { differenceInMonths, addMonths, addYears } from "date-fns/esm"

// Specify limits for the calendar in the Me tab
export const MIN_DATE = new Date(2020, 0, 1)
export const MAX_DATE = addYears(new Date(), 1)

export const getDateFromIndex = (index: number): Date =>
  addMonths(MIN_DATE, index)

export const getIndexFromDate = (date: Date): number =>
  differenceInMonths(date, MIN_DATE)
