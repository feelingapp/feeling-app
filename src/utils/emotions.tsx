import AngryIcon from "../assets/emojis/angry.svg"
import UpsetIcon from "../assets/emojis/upset.svg"
import UnsureIcon from "../assets/emojis/unsure.svg"
import GreatIcon from "../assets/emojis/great.svg"
import AmazingIcon from "../assets/emojis/amazing.svg"

export enum Emotion {
  ANGRY = "angry",
  UPSET = "upset",
  UNSURE = "unsure",
  GREAT = "great",
  AMAZING = "amazing",
}

export function getEmotionColor(emotion: Emotion): string {
  switch (emotion) {
    case Emotion.ANGRY:
      return "var(--purple-color)"
    case Emotion.UPSET:
      return "var(--blue-color)"
    case Emotion.UNSURE:
      return "var(--green-color)"
    case Emotion.GREAT:
      return "var(--yellow-color)"
    case Emotion.AMAZING:
      return "var(--red-color)"
    default:
      return "var(--light-grey-color)"
  }
}

export function renderEmoji(emotion: Emotion): JSX.Element {
  switch (emotion) {
    case Emotion.ANGRY:
      return <AngryIcon />
    case Emotion.UPSET:
      return <UpsetIcon />
    case Emotion.UNSURE:
      return <UnsureIcon />
    case Emotion.GREAT:
      return <GreatIcon />
    case Emotion.AMAZING:
      return <AmazingIcon />
    default:
      return undefined
  }
}
