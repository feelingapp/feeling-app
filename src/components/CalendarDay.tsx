import Link from "next/link"
import { styled } from "linaria/react"
import { format, isToday } from "date-fns"

import { Emotion, getEmotionColor, renderEmoji } from "../utils/emotions"

const Container = styled.a`
  display: grid;
  grid-template-areas: "content";
  justify-items: center;
  align-items: center;
  width: 100%;
`

const Background = styled.div<{ backgroundColor?: string; isToday?: boolean }>`
  grid-area: content;
  border-radius: 12px;
  background-color: ${(props) =>
    props.backgroundColor || "var(--light-grey-color)"};
  width: 100%;
  height: 0;
  padding-bottom: 100%;

  box-shadow: ${(props) =>
    props.isToday ? "0 0 0 2px rgba(0, 0, 0, 0.8) inset" : "initial"};
`

const Content = styled.div`
  grid-area: content;
  display: flex;
  flex-direction: column;
  align-items: center;

  svg {
    width: 50%;
    margin-top: var(--space-xxxs);
    filter: drop-shadow(0 0 4px rgba(0, 0, 0, 0.15));
  }
`

const Date = styled.p`
  font-weight: 600;
  font-size: clamp(10px, 2vw, 16px);
  opacity: 0.75;
`

interface Props {
  date: Date
  emotion?: Emotion
  blank?: boolean
}

export default function CalendarDay(props: Props): JSX.Element {
  // Used as an invisible placeholder when a month starts later in the week
  if (props.blank) {
    return (
      <Container as="div">
        <Background backgroundColor="transparent" />
      </Container>
    )
  }

  // Greyed out day when no feelings have been logged for that day
  if (!props.emotion) {
    return (
      <Container as="div">
        <Background isToday={isToday(props.date)} />
      </Container>
    )
  }

  return (
    <Link
      href="/me/day/[date]"
      as={"/me/day/" + format(props.date, "dd-MM-yyyy")}
      passHref={true}
    >
      <Container aria-label={format(props.date, "Mo MMMM yyyy")}>
        <Background
          backgroundColor={getEmotionColor(props.emotion)}
          isToday={isToday(props.date)}
        />
        {props.emotion && (
          <Content>
            <Date>{props.date.getDate()}</Date>
            {renderEmoji(props.emotion)}
          </Content>
        )}
      </Container>
    </Link>
  )
}
