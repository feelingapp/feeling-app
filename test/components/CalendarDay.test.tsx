import { shallow } from "enzyme"

import CalendarDay from "../../src/components/CalendarDay"
import { Emotion } from "../../src/utils/emotions"

describe("Calendar day: ", () => {
  it("blank day matches snapshot", () => {
    const wrapper = shallow(
      <CalendarDay date={new Date(2000, 1, 1)} blank={true} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it("greyed out day matches snapshot", () => {
    const wrapper = shallow(<CalendarDay date={new Date(2000, 1, 1)} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("matches snapshot", () => {
    const wrapper = shallow(
      <CalendarDay date={new Date(2000, 1, 1)} emotion={Emotion.AMAZING} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
