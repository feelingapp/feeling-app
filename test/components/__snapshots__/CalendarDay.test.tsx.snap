// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar day:  blank day matches snapshot 1`] = `
<Container
  as="div"
>
  <Background
    backgroundColor="transparent"
  />
</Container>
`;

exports[`Calendar day:  greyed out day matches snapshot 1`] = `
<Container
  as="div"
>
  <Background
    isToday={false}
  />
</Container>
`;

exports[`Calendar day:  matches snapshot 1`] = `
<Container
  as="div"
>
  <Background
    isToday={false}
  />
</Container>
`;
